goosebit/api/devices/routes.py

```diff
@@ -1,27 +1,38 @@
-import time
-from typing import Any
+from __future__ import annotations
 
 from fastapi import APIRouter, Security
 from fastapi.requests import Request
-from pydantic import BaseModel
 from tortoise.expressions import Q
 
-from goosebit.api.helper import filter_data
+from goosebit.api.responses import StatusResponse
 from goosebit.auth import validate_user_permissions
 from goosebit.models import Device, Firmware, UpdateModeEnum, UpdateStateEnum
 from goosebit.permissions import Permissions
 from goosebit.updater.manager import delete_devices, get_update_manager
 
-router = APIRouter(prefix="/devices")
+from .requests import (
+    DeleteDevicesRequest,
+    ForceUpdateDevicesRequest,
+    UpdateDevicesRequest,
+)
+from .responses import DeviceAllResponse, DeviceTableResponse, LogsDeviceResponse
+
+router = APIRouter(prefix="/devices", tags=["devices"])
 
 
 @router.get(
     "/all",
     dependencies=[Security(validate_user_permissions, scopes=[Permissions.HOME.READ])],
 )
-async def devices_get_all(request: Request) -> dict[str, int | list[Any] | Any]:
-    query = Device.all().prefetch_related("assigned_firmware", "hardware")
+async def devices_get_all(_: Request) -> DeviceAllResponse:
+    return await DeviceAllResponse.parse(await Device.all().prefetch_related("assigned_firmware", "hardware"))
+
 
+@router.get(
+    "/table",
```

Comment:
Maybe if we were to do this we would have to check if the `/devices` endpoint contains a query string?  This doesn't look very good for API documentation though, as we want a separate endpoint for the table and all response since they should contain different data ideally (or at least not contain all the table response stuff).  The other nice thing is then we can actually model the query parameters as a pydantic model so that it can be parsed automatically.

Side note fun fact, `REST` is not technically correct, heres an article - https://htmx.org/essays/how-did-rest-come-to-mean-the-opposite-of-rest/

Reply:
I'm not opposed to moving to a "RESTful" setup, but I think we need to figure out how to handle the table/all endpoint, as I dont think the way the table queries different data is the way we will want API users to be able to query, as it is very complicated.

Reply:
A general observation: some solutions opt for two APIs

1. a BFF (backend-for-frontend) that specifically caters to the FE needs.
2. an integration API for any other consumer (usually versioned with the promise not to do breaking changes)

For the BFF an upfront clean modeling is not critical as it can be changed at any time. For an integration API I would stick to RESTful to prevent surprises to the external consumer.

So maybe the current work just focusses on BFF and we add an Integration API at a later stage? But then: having two APIs is more work as well.

Or: the API provides all the RESTful endpoints. Where the UI has different needs, we create additional endpoints. Like for delete:

- DELETE /devices/{dev_id} : the expected endpoint
- DELETE /devices/bulk_delete : our BFF endpoint that has the IDs in the payload

Reply:
I like the idea of seperate endpoints, but that makes maintaining it harder.  Up to you.

Reply:
So to check if we have the same understanding. For devices, we would have the following endpoints:

- `GET /devices/{dev_id}`: returning single device, used by test code, replaces the `GET /devices/all`
- `GET /devices/table`: returning a subset of the devices as specified by the query params. Used in the UI
- `PUT /devices/bulk_update`: allows to change multiple devices at once. IDs and changed properties posted in the body, replaces `POST /devices/update` (and preferably also `POST /devices/force_update`)
- `GET /devices/{dev_id}/log`: replaces `GET /devices/logs/{dev_id}`
- `DELETE /devices/bulk_delete`: allows to delete multiple devices, IDs in body. Replaces `POST /devices/delete`

Reply:
I think where we are at now I can start splitting them, but heres my thoughts - 

API: 
Prefix: `/api/v1`
- `GET /devices` (replaces `GET /devices/all`)
- `GET /devices/{dev_id}`
- `GET /devices/{dev_id}/log` (as above)
- `PATCH /devices` (bulk update)
- `PATCH /devices/{dev_id}` (single device update)
- `DELETE /devices` (bulk delete)
- `DELETE /devices/{dev_id}` (single device delete)

BFF:
Prefix: `/api/frontend`
- `GET /devices` (replaces `GET /devices/table`)
- `DELETE /devices` (bulk delete, will directly call into API function)
- Any other endpoints we want to copy from the API, should mostly be similar.

Reply:
OK, I see. But why not move the special cases (the once that are not RESTful) to BFF? And also: `GET /devices/all` is not needed. The test code extracts a single device from this call, so `GET /devices/{dev_id}` would be more convenient.

API:
Prefix: `/api/v1`
- `GET /devices/{dev_id}`
- `GET /devices/{dev_id}/log` (as above)
- `PATCH /devices/{dev_id}` (single device update)
- `DELETE /devices/{dev_id}` (single device delete)

BFF:
Prefix: `/api/frontend`
- `GET /devices` (replaces GET /devices/table)
- `PATCH /devices` (bulk update)
- `DELETE /devices` (bulk delete)

Reply:
Thinking a bit more about it: maybe UI should only use /frontend. Could result in duplicating a few endpoints but that should not be an issue.

Reply:
I actually prefer `/api/bff`? Pun intended 😉

Reply:
> And also: GET /devices/all is not needed. The test code extracts a single device from this call, so GET /devices/{dev_id} would be more convenient.

This is mainly for end users, you can imagine a user would want to query for ALL connected devices.  We may want to have a way for a use to just query a list of UUIDs, but it might be more useful to have all the information in one place.

> I actually prefer /api/bff. Pun intended 😉

So I think then we should decide if we are all in agreement about the prefixes.  Seems like there is little to no debate on the endpoints (plus that can be debated when those changes are made if needed), but I just want to confirm the thinking there -

`/api/v1` - version 1 of the API.
`/api/v{version}` version {version} of the API.
`/api/bff` internal backend-for-frontend API.  Do not include in `/docs` or schema.

> Thinking a bit more about it: maybe UI should only use /frontend. Could result in duplicating a few endpoints but that should not be an issue.

Keep in mind that since these are all just python functions, technically an endpoint like `/api/bff/devices` could look like this (simplified):


```
from goosebit.api.devices.routes import get_devices

@router.get("/devices")
async def bff_get_devices(request: Request):
    return await get_devices(request)
```

Reply:
> bff_get_devices

Can‘t the module be called `bff` and the function just `get_devices`?

Reply:
> > And also: GET /devices/all is not needed. The test code extracts a single device from this call, so GET /devices/{dev_id} would be more convenient.
> 
> This is mainly for end users, you can imagine a user would want to query for ALL connected devices. We may want to have a way for a use to just query a list of UUIDs, but it might be more useful to have all the information in one place.

OK, but, in REST that would normally be returned by a GET on `/api/v1/devices`.

Reply:
> OK, but, in REST that would normally be returned by a GET on /api/v1/devices, which you actually suggested.

Yep, this is the plan.

> Can‘t the module be called bff and the function just get_devices?

It was just an example to clarify, and I wanted to specific which was which.

Reply:
Agree that we should understand the prefixes (als relevant for #66). Here the status quo how I understand it:
 
- `/api/v1`: RESTful stable integration API. Ideally supports API keys for authentication?
- `/api/bff`: API for the frontend, currently uses sessions for authentication.
- `/ui`: to get the static html ui pages, currently uses sessions for authentication.
- `/{tenant}/controller/v1`: DDI, currently no authentication, but should support custom header (generated by trusted service based on device certificate) used to validate dev_id in path (ie. ensure, that devices can only report things for themselves). Note: {tenant} should not be set to the value ui or api.

Not under any prefix (yet) are `/login`, `/logout` and `/download`.

Should it be the following?
- `/ddi/controller/v1`: we don't plan for tenant support, so this keeps it simple. I'd consider to move the download path in there as well (as /ddi/controller/v1/{dev_id}/download) – same as hawkBit. Not helpful if CDN should be used - but this use case is covered by gooseBit through the "remote firmwares".
-  `/api/v1`: RESTful stable integration API v1. Playing with `/mgmt/v1`or something similar to make it more consistent with ddi and ui. But don't like mgmt much and fail to find something better.
- `/ui`: to contain everything ui related (so also login/logout and the bff path). 

With this we would group the endpoints by their authentication mechanism (`/ui/login` being the only special case)

Reply:
> `/ddi/controller/v1`: we don't plan for tenant support, so this keeps it simple. I'd consider to move the download path in there as well (as /ddi/controller/v1/{dev_id}/download) – same as hawkBit. Not helpful if CDN should be used - but this use case is covered by gooseBit through the "remote firmwares".

Agreed. I think moving download here makes sense, but my only concern is downloading from the UI.  Do we want to have an API endpoint to download as well?  I think they will all just end up calling each other anyway, but I think we need essentially 2 of the same endpoint, one in `bff` and one in `ddi`.

> With this we would group the endpoints by their authentication mechanism (`/ui/login` being the only special case)

Biggest issue I have with moving the login endpoints right now is that the API router used for API and UI both require auth at the top level, and I would prefer to keep them that way.  I don't like the merging of multiple base level routers into one router (like what was done with API to handle the download endpoint originally), so it might be preferable to have just a static endpoint mounted into the FastAPI instance at `/api/v1/login` and `/ui/login`?

Reply:
I think I resolved this for now, however I havent moved the login endpoint.  I want to maybe do that in a separate PR if needed, as this set of changes is already becoming quite large.